core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,305 @@
+    pub fn root(&self) -> Word {
+        self.root
+    }
+
+    /// Returns the depth of this Merkle tree.
+    ///
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        self.total_depth
+    }
+
+    /// Returns a node at the specified index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_node(&self, index: u64) -> Result<Word, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+
+        let pos = 2u64.pow(self.total_depth) + index;
+        let index = (pos / 2) as u64;
+        if is_even(pos) {
+            Ok(self.paths[&index][0])
+        } else {
+            Ok(self.paths[&index][1])
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_path(&self, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+
+        let pos = 2u64.pow(self.total_depth) + index;
+        let index = pos / 2;
+
+        let mut path = self.paths[&index].clone();
+
+        if is_even(pos) {
+            path.remove(0);
+            Ok(path)
+        } else {
+            path.remove(1);
+            Ok(path)
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.total_depth;
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+
+        let mut pos = 2u64.pow(depth) + index;
+
+        // Fill old_hashes vector -----------------------------------------------------------------
+        let path = self.paths.get(&(pos / 2)).unwrap();
+        let mut path_iter = path.iter();
+
+        let mut old_hashes = Vec::<Word>::new();
+
+        let mut path_hash;
+        let mut old_hash;
+
+        if is_even(pos) {
+            old_hash = *path_iter.next().unwrap();
+            path_hash = *path_iter.next().unwrap();
+        } else {
+            path_hash = *path_iter.next().unwrap();
+            old_hash = *path_iter.next().unwrap();
+        }
+
+        for _ in 0..depth - 2 {
+            old_hash = calculate_parent_hash(old_hash, pos, path_hash);
+            pos /= 2;
+            old_hashes.push(old_hash);
+            path_hash = *path_iter.next().unwrap();
+        }
```

Comment:
I think what you are doing here is building a list of old values of nodes in the tree which will change as the result of changing the leaf, and in the next section you actually compute these new nodes. If my understanding is correct, I'd create a helper function which looks like this:

```Rust
fn compute_hashes(path: &[Word], index: usize) -> Vec<Word>
```

As `index` you can always pass the index of the even node and I think the result should be the same. The returned vector will have 1 fewer element than the input path because the first two values in the passed-in path are leaves which will get merged into a single node.